packages/tracing/src/hubextensions.ts

```diff
@@ -19,25 +20,68 @@ function traceHeaders(this: Hub): { [key: string]: string } {
+ * decision, which all child spans and child transactions inherit.
+ *
+ * Sample rate is set in SDK config, either as a constant (`tracesSampleRate`) or a function to compute the rate
+ * (`tracesSampler`).
+ *
+ * Called every time a transaction is created. Only transactions which emerge with sampled === true will be sent to
+ * Sentry.
+ *
+ * Mutates the given Transaction object and then returns the mutated object.
  */
 function sample<T extends Transaction>(hub: Hub, transaction: T): T {
   const client = hub.getClient();
+  const options = (client && client.getOptions()) || {};
+
+  // nothing to do if tracing is disabled
+  if (!hasTracingEnabled(options)) {
+    transaction.sampled = false;
+    return transaction;
+  }
+
+  logger.log('Tracing enabled');
+
+  // we have to test for a pre-existsing sampling decision, in case this transaction is a child transaction and has
+  // inherited its parent's decision
   if (transaction.sampled === undefined) {
-    const sampleRate = (client && client.getOptions().tracesSampleRate) || 0;
-    // if true = we want to have the transaction
-    // if false = we don't want to have it
-    // Math.random (inclusive of 0, but not 1)
+    let sampleRate;
+
+    // prefer the hook
+    if (options.tracesSampler) {
+      // TODO (kmclb) build context object
+      const sampleContext: SampleContext = {};
+      sampleRate = options.tracesSampler(sampleContext);
+    }
+    // we would have bailed at the beginning if neither `tracesSampler` nor `tracesSampleRate` were defined, so if the
+    // former isn't, the latter must be
+    else {
+      sampleRate = options.tracesSampleRate;
+    }
+
+    // if the function returned either 0 or null, it's a sign the transaction should be dropped
+    if (!sampleRate) {
+      logger.log('Discarding trace because tracesSampler returned 0 or null');
+      transaction.sampled = false;
+      return transaction;
+    }
+
+    // now we roll the dice (Math.random is inclusive of 0, but not of 1)
     transaction.sampled = Math.random() < sampleRate;
-  }
 
-  // We only want to create a span list if we sampled the transaction
-  // If sampled == false, we will discard the span anyway, so we can save memory by not storing child spans
-  if (transaction.sampled) {
-    const experimentsOptions = (client && client.getOptions()._experiments) || {};
-    transaction.initSpanRecorder(experimentsOptions.maxSpans as number);
+    // if we're not going to keep it, we're done
+    if (!transaction.sampled) {
+      logger.log(`Discarding trace because it's not included in the random sample (sampling rate = ${sampleRate})`);
+      return transaction;
+    }
   }
 
+  // at this point we know we're keeping the transaction, whether because of an inherited decision or because it got
+  // lucky with the dice roll
+  const experimentsOptions = (client && client.getOptions()._experiments) || {};
```

Comment:
If my patch above will be applied, then this can be changed to `options._experiments || {}`

Reply:
You're right. Actually, even without your patch the check was unnecessary. Fixed.